File: Algo_Ds_Notes/Tim_Sort/Tim_Sort.py
#GSSoC
#TimSort 


# MERGE SORT FUNCTION :
def merge(array, left, mid, right): 

    leftLength, rightLength = mid - left + 1, right - mid 
    Left, Right = [], [] 
    for i in range(0, leftLength): 
        Left.append(array[left + i]) 
    for i in range(0, rightLength): 
        Right.append(array[mid + 1 + i]) 
	
    i, j, k = 0, 0, left 

    while i < leftLength and j < rightLength: 
	
        if Left[i] <= Right[j]: 
            array[k] = Left[i] 
            i += 1
		
        else: 
            array[k] = Right[j] 
            j += 1
		
        k += 1
	
    while i < leftLength: 
	
        array[k] = Left[i] 
        k += 1
        i += 1
	
    while j < rightLength: 
        array[k] = Right[j] 
        k += 1
        j += 1


# INSERTION SORT FUNCTION :
def insertionSort(array, left, right): 

    for i in range(left + 1, right+1): 
	
        temp = array[i] 
        j = i - 1
        while array[j] > temp and j >= left: 
		
            array[j+1] = array[j] 
            j -= 1
		
        array[j+1] = temp
	

# TIMSORT FUNCTION :
def timSort(array, n): 
    Limit=32 
    for i in range(0, n, Limit): 
        insertionSort(array, i, min((i+31), (n-1))) 
    size = Limit 
    while size < n: 
	
        for left in range(0, n, 2*size): 
		 
            mid = left + size - 1
            right = min((left + 2*size - 1), (n-1)) 
	
            merge(array, left, mid, right) 
		
        size = 2*size 
		
	
if __name__ == "__main__": 

    array = [50, 39, 89, 10, 8] 
    n = len(array) 
    print("Unsorted Array :") 
    for i in range(0, n): 
        print(array[i], end = " ") 
    print()
	
    timSort(array, n) 
	
    print("Sorted Array :") 
    for i in range(0, n): 
        print(array[i], end = " ") 
    print() 
	

'''
OUTPUT
----------------
Unsorted Array :
50 39 89 10 8 
Sorted Array :
8 10 39 50 89
----------------
'''
# Happy Coding!

File: Tim_Sort/Tim_Sort.php
<?php
function insertionSort(&$arr,$left,$right)
{
  for($i=$left+1;$i<=$right;$i++)
  {
        $temp = $arr[$i]; 
        $j = $i-1; 
        while ($j>=$left && $arr[$j] > $temp) 
        { 
            $arr[$j+1] = $arr[$j]; 
            $j--; 
        } 
        $arr[$j+1] = $temp; 
    } 
}
// merge function merges the sorted runs 
function merge(&$arr, $l, $m, $r) 
{ 
    // original array is broken in two parts 
    // left and right array 
    $len1 = $m - $l + 1;
    $len2 = $r - $m;
    $left = new SplFixedArray($len1);
    $right = new SplFixedArray($len2);
    for ($i = 0; $i < $len1; $i++) 
        $left[$i] = $arr[$l + $i]; 
    for ($i = 0; $i < $len2; $i++) 
        $right[$i] = $arr[$m + 1 + $i]; 
  
    $i = 0; 
    $j = 0; 
    $k = $l; 
  
    // after comparing, we merge those two array 
    // in larger sub array 
    while ($i < $len1 && $j <$len2) 
    { 
        if ($left[$i] <= $right[$j]) 
        { 
            $arr[$k] = $left[$i]; 
            $i++; 
        } 
        else
        { 
            $arr[$k] = $right[$j]; 
            $j++; 
        } 
        $k++; 
    }
    // copy remaining elements of left, if any 
    while ($i < $len1) 
    { 
        $arr[$k] = $left[$i]; 
        $k++; 
        $i++; 
    }
    // copy remaining element of right, if any 
    while ($j < $len2) 
    { 
        $arr[$k] = $right[$j]; 
        $k++; 
        $j++; 
    } 
}
// iterative Timsort function to sort the 
// array[0...n-1] (similar to merge sort) 
function timSort(&$arr) 
{ 
  $run=32;
    // Sort individual subarrays of size RUN 
    for ($i = 0; $i < count($arr); $i+=$run) 
        insertionSort($arr, $i,min(($i+31), (count($arr)-1))); 
    // start merging from size RUN (or 32). It will merge 
    // to form size 64, then 128, 256 and so on .... 
    for ($size = $run; $size < count($arr); $size = 2*$size) 
    { 
        // pick starting point of left sub array. We 
        // are going to merge arr[left..left+size-1] 
        // and arr[left+size, left+2*size-1] 
        // After every merge, we increase left by 2*size 
        for ($left = 0; $left < count($arr); $left += 2*$size) 
        { 
            // find ending point of left sub array 
            // mid+1 is starting point of right sub array 
            $mid = $left + $size - 1; 
            $right = min(($left + 2*$size - 1), (count($arr)-1)); 
  
            // merge sub array arr[left.....mid] & 
            // arr[mid+1....right] 
            merge($arr, $left, $mid, $right); 
        } 
    } 
} 
  
//function to print the Array 
function printArray($arr) 
{ 
    $output = "[";
    for ($i = 0; $i < count($arr); $i++){
        $output .= " ".$arr[$i]." "; 
        if ($i<count($arr)-1) {
            $output .=",";
        }
    }
    $output .= "]";
    echo $output;
} 


$arr = array(9,65,0,2,-4,12,-18); 
print("\nGiven Array is\n"); 
printArray($arr); 
timSort($arr); 
print("\nAfter Sorting Array is\n"); 
printArray($arr); 

// OUTPUT =>
// ------------
// Given Array is
//[ 9 , 65 , 0 , 2 , -4 , 12 , -18 ]
//After sorting Array is
//[ -18 , -4 , 0 , 2 , 9 , 12 , 65 ]
?>